src/dataunion/DataUnion.ts

```diff
@@ -564,30 +590,109 @@ export class DataUnion {
-        getWithdrawTxFunc: () => Promise<TransactionResponse>,
+        getWithdrawTxFunc: () => Promise<ContractTransaction>,
         recipientAddress: EthereumAddress,
         options: DataUnionWithdrawOptions = {}
-    ): Promise<TransactionReceipt> {
+    ): Promise<ContractReceipt | AmbMessageHash | null> {
         const {
             pollingIntervalMs = 1000,
-            retryTimeoutMs = 60000,
-            freeWithdraw = this.client.options.dataUnion.freeWithdraw,
+            retryTimeoutMs = 300000,
+            // by default, transport the signatures if freeWithdraw isn't supported by the sidechain
+            payForTransport = !this.client.options.dataUnion.freeWithdraw,
+            waitUntilTransportIsComplete = true,
             sendToMainnet = true,
         } = options
+
         const getBalanceFunc = sendToMainnet
             ? () => this.client.getTokenBalance(recipientAddress)
             : () => this.client.getSidechainTokenBalance(recipientAddress)
-        const balanceBefore = await getBalanceFunc()
+        const balanceBefore = waitUntilTransportIsComplete ? await getBalanceFunc() : 0
+
+        log('Executing DataUnionSidechain withdraw function')
         const tx = await getWithdrawTxFunc()
         const tr = await tx.wait()
-        if (!freeWithdraw && sendToMainnet) {
-            await this.getContracts().transportSignaturesForTransaction(tr, options)
+
+        // keep tokens in the sidechain => just return the sidechain tx receipt
+        if (!sendToMainnet) { return tr }
+
+        log(`Got receipt, filtering UserRequestForSignature from ${tr.events!.length} events...`)
+        const ambHashes = getMessageHashes(tr)
+
+        if (ambHashes.length < 1) {
+            throw new Error("No UserRequestForSignature events emitted from withdraw transaction, can't transport withdraw to mainnet")
+        }
+
+        if (ambHashes.length > 1) {
+            throw new Error('Expected only one UserRequestForSignature event')
+        }
+
+        const messageHash = ambHashes[0]
+
+        if (!payForTransport) {
+            // expect someone else to do the transport for us (corresponds to dataUnion.freeWithdraw=true)
+            if (waitUntilTransportIsComplete) {
+                log(`Waiting for balance=${balanceBefore.toString()} change (poll every ${pollingIntervalMs}ms, timeout after ${retryTimeoutMs}ms)`)
+                await until(async () => !(await getBalanceFunc()).eq(balanceBefore), retryTimeoutMs, pollingIntervalMs).catch((e) => {
+                    const msg = `Timeout: Bridge did not transport withdraw message as expected. Fix: DataUnion.transportMessage("${messageHash}")`
+                    throw e.message.startsWith('Timeout') ? new Error(msg) : e
+                })
+                return null
+            }
+
+            // instead of waiting, hand out the messageHash so that someone else might do the transport using it
+            return messageHash
```

Comment:
This return type is a bit weird `ContractReceipt | AmbMessageHash | null`, I wonder if instead of a `payForTransport` flag we should have two methods? Might be a bit of work though.

Reply:
those are kind of... special cases of the same thing, somehow. Yes the type is super weird, but I guess the other solution would be to rewrite it so that it's exported in smaller pieces so that if you don't want the "whole thing" you can just take the pieces you want, or something. Though that's how it is now, too, more or less.